omnigibson/utils/teleop_utils.py

```diff
@@ -0,0 +1,540 @@
+            elif device.get_class() == self.xr_device_class.xrtracker:
+                self.trackers[device.get_index()] = device
+        assert self.hmd is not None, "[VRSys] HMD not detected! Please make sure you have a VR headset connected to your computer."
+
+    def _update_device_transforms(self) -> None:
+        """
+        Get the transform matrix of each VR device *in world frame* and store in self.raw_data
+        Note that we have to transpose the transform matrix because Omniverse uses row-major matrices while OmniGibson uses column-major matrices
+        """
+        transforms = {}
+        transforms["hmd"] = self.xr2og(self.hmd.get_virtual_world_pose())
+        transforms["controllers"] = {}
+        transforms["trackers"] = {}
+        for controller_index in self.controllers:
+            transforms["controllers"][controller_index] = self.xr2og(self.controllers[controller_index].get_virtual_world_pose())
+        for tracker_index in self.trackers:
+            transforms["trackers"][tracker_index] = self.xr2og(self.trackers[tracker_index].get_virtual_world_pose())
+        self.raw_data["transforms"] = transforms
+
+    def _update_button_data(self):
+        """
+        Get the button data for each controller and store in self.raw_data
+        Returns:
+            dict: a dictionary of whether each button is pressed or touched, and the axis state for touchpad and joysticks
+        """
+        button_data = {}
+        for controller_index in self.controllers:
+            button_data[controller_index] = {}
+            button_data[controller_index]["press"] = self.controllers[controller_index].get_button_press_state()
+            button_data[controller_index]["touch"] = self.controllers[controller_index].get_button_touch_state()
+            button_data[controller_index]["axis"] = self.controllers[controller_index].get_axis_state()
+        self.raw_data["button_data"] = button_data       
+    
+    def _update_hand_tracking_data(self, e: carb.events.IEvent) -> None:
+        """
+        Get hand tracking data, see https://registry.khronos.org/OpenXR/specs/1.0/html/xrspec.html#convention-of-hand-joints for joint indices
+        Args:
+            e (carb.events.IEvent): event that contains hand tracking data as payload
+        """
+        e.consume()
+        data_dict = e.payload
+        for hand in ["left", "right"]:
+            self.raw_data["hand_data"][hand] = {}
+            if data_dict[f"joint_count_{hand}"] != 0:
+                self.raw_data["hand_data"][hand]["raw"] = {"pos": [], "orn": []}
+                hand_joint_matrices = data_dict[f"joint_matrices_{hand}"]
+                for i in range(26):
+                    pos, orn = self.xr2og(np.reshape(hand_joint_matrices[16 * i: 16 * (i + 1)], (4, 4)))
+                    self.raw_data["hand_data"][hand]["raw"]["pos"].append(pos)
+                    self.raw_data["hand_data"][hand]["raw"]["orn"].append(orn)
```

Comment:
hmm maybe at least add comments explaining why 16 and 26?

Reply:
done